From e05d193068b6032d5aec24f9716489e290a5e47a Mon Sep 17 00:00:00 2001
From: chdugw_auto <jkulkarni@maxlinear.com>
Date: Tue, 31 Aug 2021 13:53:56 +0300
Subject: [PATCH] Merge pull request #3501 in SW_UGW/linux from
 feature/PONRTSYS-9417-proper-way-of-handling-queue-map-replication to master

* commit 'dab3fa0555c2f3e6d227218c6612c337b865a25d':
  PONRTSYS-9417: When deleting tc to queue mappings, map to default queue
---
 drivers/net/datapath/pon_qos/pon_qos_tc_qos.c | 33 ++++++++++++++++++++++-----
 1 file changed, 27 insertions(+), 6 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
index a2a4c57bfdcf..1126d7d824da 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_qos.c
@@ -819,14 +819,14 @@ int pon_qos_tc_sched_status(struct pon_qos_port *p, struct pon_qos_qdisc *root)
 	return 0;
 }
 
-static int pon_qos_set_qmap(struct pon_qos_q_data *qid, int port, int subif,
+static int pon_qos_set_qmap(u32 qid, int port, int subif,
 			    int tc_cookie, bool en)
 {
 	struct dp_queue_map_set qmap_set = {0};
 	int ret;
 
 	qmap_set.inst = 0;
-	qmap_set.q_id = en ? qid->qid : 0;
+	qmap_set.q_id = en ? qid : 0;
 
 	qmap_set.map.dp_port = port;
 	qmap_set.map.subif = subif;
@@ -862,7 +862,7 @@ static int pon_qos_config_cpu_port(struct pon_qos_q_data *qid,
 {
 	int ret;
 
-	ret = pon_qos_set_qmap(qid, port, subif, tc_cookie, en);
+	ret = pon_qos_set_qmap(qid->qid, port, subif, tc_cookie, en);
 	if (ret != 0)
 		return ret;
 
@@ -952,7 +952,7 @@ static int pon_qos_map_qid(struct pon_qos_q_data *qid,
 {
 	int ret;
 
-	ret = pon_qos_set_qmap(qid, sch->port, subif, tc_cookie, en);
+	ret = pon_qos_set_qmap(qid->qid, sch->port, subif, tc_cookie, en);
 	if (ret != 0)
 		return -EFAULT;
 
@@ -996,7 +996,22 @@ static int pon_qos_replicate_mappings_for_subifs(struct net_device *dev,
 		if (subif.port_id != sch->port)
 			continue;
 
-		ret = pon_qos_set_qmap(qid, sch->port, subif.subif, tc, en);
+		/*
+		 * If en == false, this should restore the default queue
+		 * mappings. If en is false in pon_qos_set_qmap() we will map
+		 * to qid=0 which is the drop queue.
+		 * So when we stop the omcid the normal traffic is dropped, but
+		 * the default setting done by the eth driver is to make the
+		 * traffic pass by default.
+		 * Therefore we should map to default queue, instead of drop
+		 * queue, if en == false.
+		 */
+		if (en)
+			ret = pon_qos_set_qmap(qid->qid, sch->port, subif.subif,
+					       tc, en);
+		else
+			ret = pon_qos_set_qmap(sch->def_q, sch->port,
+					       subif.subif, tc, true);
 		if (ret)
 			goto cleanup;
 		i++;
@@ -1004,6 +1019,7 @@ static int pon_qos_replicate_mappings_for_subifs(struct net_device *dev,
 
 cleanup:
 	if (ret) {
+		en = !en;
 		for_each_netdev(&init_net, netdev) {
 			if (dp_get_netif_subifid(netdev, NULL, NULL, NULL, &subif, 0))
 				continue;
@@ -1012,7 +1028,12 @@ static int pon_qos_replicate_mappings_for_subifs(struct net_device *dev,
 			if (i <= 0)
 				break;
 
-			pon_qos_set_qmap(qid, sch->port, subif.subif, tc, !en);
+			if (en)
+				ret = pon_qos_set_qmap(qid->qid, sch->port,
+						       subif.subif, tc, en);
+			else
+				ret = pon_qos_set_qmap(sch->def_q, sch->port,
+						       subif.subif, tc, true);
 			i--;
 		}
 	}
